### ISSUE TITLE
NaN FloatDomainError when using the :eta marker in v0.18.0

### ISSUE THREAD

**Issue by DannyBen** [piotrmurach/tty-progressbar#48]
### Describe the problem

After upgrading from 0.17.0 to 0.18.0, I started receiving `NaN (FloatDomainError)` exceptions.
I tracked it to the `:eta` marker.

### Steps to reproduce the problem

Save this self-contained ruby file and run it.

```ruby
require 'bundler/inline'

gemfile do
  source "https://rubygems.org"
  
  gem 'tty-progressbar', '0.18.0'     # BROKEN
  # gem 'tty-progressbar', '0.17.0'   # OK
end

def progress_bar(text, count)
  # BROKEN
  TTY::ProgressBar.new "#{text} :bar ETA :eta", total: count

  # OK
  # TTY::ProgressBar.new "#{text} :bar", total: count
end

total = 50
bar = progress_bar "downloading", total
total.times do |i|
  bar.log "Working on #{i}/#{total}"
  sleep 0.1
  bar.advance
end
```

### Actual behaviour

Exception raised:

```
/store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar/converter.rb:18:in `floor': NaN (FloatDomainError)
        from /store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar/converter.rb:18:in `to_time'
        from /store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar/formatter/estimated.rb:28:in `call'
        from /store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar/pipeline.rb:44:in `block in decorate'
        from /store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar/pipeline.rb:42:in `each'
        from /store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar/pipeline.rb:42:in `inject'
        from /store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar/pipeline.rb:42:in `decorate'
        from /usr/share/rvm/rubies/ruby-3.0.0/lib/ruby/3.0.0/forwardable.rb:238:in `decorate'
        from /store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar.rb:345:in `render'
        from /store/gems/ruby-3.0.0/gems/tty-progressbar-0.18.0/lib/tty/progressbar.rb:555:in `log'
        from test.rb:20:in `block in <main>'
        from test.rb:19:in `times'
        from test.rb:19:in `<main>'

```

### Expected behaviour

It is expected to work as version 0.17.0, not crash and respect the `:eta` marker.

### Describe your environment

* OS version: 18.04.5 LTS (Bionic Beaver)
* Ruby version: ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]
* TTY::ProgressBar version: 0.18.0

**Comment by piotrmurach** [comment]
Thank you for this report!

Initially, I couldn't believe it that the `:eta` token is broken as I have tested it extensively running automated tests and manually all examples like [pause](https://github.com/piotrmurach/tty-progressbar/blob/master/examples/pause.rb) or [indeterminate](https://github.com/piotrmurach/tty-progressbar/blob/master/examples/indeterminate.rb) on different operating systems. Thankfully, it still works fine in these examples albeit with a caveat that I will explain.

In short, the culprit turned out to be the `log` method or perhaps I should say that thanks to the new `v0.18.0` release, a bug that was always there become visible.

For example, if you remove `log` method, call `start` before the loop or swap `log` and `advance` calls your snippet will work fine:

```ruby
total.times do |i|
  bar.advance
  sleep 0.1
  bar.log "Working on #{i}/#{total}"
end
```

The bug is to do with how measuring start time happens which is then used for eta calculation. Generally, an internal timer can be started either by `start` method or automatically when `advance` is called for the first time.

 In the old `v0.17.0`, when you initialize a progress bar, a start time gets initialized to `Time.now` and then when the `advance` is called for the first time, this start time is overridden to be the actual start time. And here is the catch, when you call `log` method, this causes a render of the progress bar, which in turn evaluates all the tokens including `:eta`. But neither `start` or `advance` have been called yet. At this point, the `:eta` token is calculated as `Time.now - start_time` and then divided by progress ratio. The ratio is zero. So the division would be, for example, ```12 / 0.0``` which resulted in `Infinity` and not an error. The `Infinity` was handled and the eta was calculated as 0.

In the new `v0.18.0`, the internal timer is a new class with the ability to start and stop. This improved time tracking significantly. Made is also easier. One benefit is that the timer gets **only** started when `advance` or `start` method is called. So in this case, when `log` method is called, the `timer.elapsed_time` returns 0. The ratio is zero. Boom! A nice error.

This may be more background than necessary, but I found it interesting tracking the bug and thought I will share my findings.

Now, the question needs answering is "What should be displayed when `log` is called before `advance` or `start`?" 
One option is to check the ratio for when it's zero and show `0s` for eta and `00:00:00` for eta time. The other option is to show `-s` and `--:--:--` similar to indeterminate progress as at this point we really don't know what the eta can be so showing zero doesn't seem to be accurate. What's your thought on this?

**Comment by DannyBen** [comment]
Thanks for sharing the interesting explanation.

I didn't realize that it was the order of my calls that "sneaked one" past the unit test goalie.

In my opinion, anything that is yet unknown (ETA / ETA Time) should either be blank (spaces, or `''`, depending on the underlying constraints), or as you suggested, something that denotes "unknown", like `--:--:--`. The other option of `00:00:00` is also reasonable (since we are used to it from clocks after an electricity outage), but it is only my third choice our of the three.

**Comment by piotrmurach** [comment]
This should be fixed now - see https://github.com/piotrmurach/tty-progressbar/commit/0f971278679e32363969b33aab4e07ba807dd000. There was a small catch to do with manual timer start using `start` method. I believe the `:eta` and `:eta_time` tokens should work now in any situation. I also updated [log example](https://github.com/piotrmurach/tty-progressbar/blob/master/examples/log.rb). There is no way this will be broken going forward. Do you have time to check it out?

**Comment by DannyBen** [comment]
> Do you have time to check it out?

Sure. Checking.

**Comment by DannyBen** [comment]
Tested both the above test script and my actual code that uses it with the `master` version, all seems to be in order.

**Comment by piotrmurach** [comment]
Released `v0.18.1`.
